datameta/api/users.py

```diff
@@ -0,0 +1,107 @@
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from pyramid.view import view_config
+from pyramid.request import Request
+from pyramid.httpexceptions import HTTPNoContent, HTTPForbidden
+
+from dataclasses import dataclass
+from . import DataHolderBase
+from ..models import User, Group
+from .. import security, errors
+from ..resource import resource_by_id
+
+@dataclass
+class UserUpdateRequest(DataHolderBase):
+    """Class for User Update Request communication to OpenApi"""
+    name: str
+    groupId: str
+    groupAdmin: bool
+    siteAdmin: bool
+    enabled: bool
+
+@view_config(
+    route_name="user_id", 
+    renderer='json', 
+    request_method="PUT", 
+    openapi=True
+)
+def get(request: Request) -> UserUpdateRequest:
+    """Change the name of the group"""
+
+    user_id = request.matchdict["id"]
+    name = request.openapi_validated.body.get("name")
+    group_id = request.openapi_validated.body.get("groupId")
+    group_admin = request.openapi_validated.body.get("groupAdmin")
+    site_admin = request.openapi_validated.body.get("siteAdmin")
+    enabled = request.openapi_validated.body.get("enabled")
+
+    db = request.dbsession
+
+    # Authenticate the user
+    auth_user = security.revalidate_user(request)
+    
+    # Get the targeted user
+    target_user = resource_by_id(db, User, user_id)
+
+    # First, check, if the user has the rights to perform all the changes they want
+
+    # The user has to be site admin to change another users group
+    if group_id != None:
+        if not auth_user.site_admin:
+            raise HTTPForbidden()
+
+    # The user has to be site admin to make another user site admin
+    if site_admin != None:
+        if not auth_user.site_admin:
+            raise HTTPForbidden()
+
+    # The user has to be site admin or group admin of the users group to make another user group admin
+    if group_admin != None:
+        if not (auth_user.site_admin or auth_user.group_admin and auth_user.group.uuid == target_user.group.uuid):
+            raise HTTPForbidden()
+
+    # The user has to be site admin or group admin of the users group to enable or disable a user
+    if enabled != None:
+        if not (auth_user.site_admin or auth_user.group_admin and auth_user.group.uuid == target_user.group.uuid):
+            raise HTTPForbidden()
+
+    # The user can change their own name or be site admin or group admin of the users group to change the name of another user
+    if name != None:
+        if not (auth_user.site_admin or auth_user.group_admin and auth_user.group.uuid == target_user.group.uuid
+            or auth_user.uuid == target_user.uuid):
+            raise HTTPForbidden()
+
+    # Now, make the corresponding changes
+    if group_id != None:
+        new_group = resource_by_id(db, Group, group_id)
```

Comment:
`resource_by_id()` returns `None` if the group_id cannot be found. Needs checking and a response.